xds/src/main/java/io/grpc/xds/LoadBalancerConfigFactory.java

```diff
@@ -0,0 +1,326 @@
+    if (minRingSize != null) {
+      configBuilder.put(MIN_RING_SIZE_FIELD_NAME, minRingSize.doubleValue());
+    }
+    if (maxRingSize != null) {
+      configBuilder.put(MAX_RING_SIZE_FIELD_NAME, maxRingSize.doubleValue());
+    }
+    return ImmutableMap.of(RING_HASH_FIELD_NAME, configBuilder.build());
+  }
+
+  /**
+   * Builds a service config JSON object for the least_request load balancer config based on the
+   * given config values..
+   */
+  private static ImmutableMap<String, ?> buildLeastRequestConfig(Integer choiceCount) {
+    ImmutableMap.Builder<String, Object> configBuilder = ImmutableMap.builder();
+    if (choiceCount != null) {
+      configBuilder.put(CHOICE_COUNT_FIELD_NAME, choiceCount.doubleValue());
+    }
+    return ImmutableMap.of(LEAST_REQUEST_FIELD_NAME, configBuilder.build());
+  }
+
+  /**
+   * Builds a service config JSON wrr_locality by wrapping another policy config.
+   */
+  private static ImmutableMap<String, ?> buildWrrLocalityConfig(
+      ImmutableMap<String, ?> childConfig) {
+    return ImmutableMap.<String, Object>builder().put(WRR_LOCALITY_FIELD_NAME,
+        ImmutableMap.of(CHILD_POLICY_FIELD, ImmutableList.of(childConfig))).build();
+  }
+
+  /**
+   * Builds an empty service config JSON config object for round robin (it is not configurable).
+   */
+  private static ImmutableMap<String, ?> buildRoundRobinConfig() {
+    return ImmutableMap.of(ROUND_ROBIN_FIELD_NAME, ImmutableMap.of());
+  }
+
+  /**
+   * Responsible for converting from a {@code envoy.config.cluster.v3.LoadBalancingPolicy} proto
+   * message to a gRPC service config format.
+   */
+  static class LoadBalancingPolicyConverter {
+
+    private static final int MAX_RECURSION = 16;
+
+    /**
+     * Converts a {@link LoadBalancingPolicy} object to a service config JSON object.
+     */
+    private static ImmutableMap<String, ?> convertToServiceConfig(
+        LoadBalancingPolicy loadBalancingPolicy, int recursionDepth)
+        throws ResourceInvalidException {
+      if (recursionDepth > MAX_RECURSION) {
+        throw new MaxRecursionReachedException();
+      }
+      ImmutableMap<String, ?> serviceConfig = null;
+
+      for (Policy policy : loadBalancingPolicy.getPoliciesList()) {
+        Any typedConfig = policy.getTypedExtensionConfig().getTypedConfig();
+        try {
+          if (typedConfig.is(RingHash.class)) {
+            serviceConfig = convertRingHashConfig(typedConfig);
+          } else if (typedConfig.is(WrrLocality.class)) {
+            serviceConfig = convertWrrLocalityConfig(typedConfig, recursionDepth);
+          } else if (typedConfig.is(RoundRobin.class)) {
+            serviceConfig = convertRoundRobinConfig();
+          } else if (typedConfig.is(TypedStruct.class)) {
+            serviceConfig = convertCustomConfig(typedConfig);
+          }
+          // TODO: support least_request once it is added to the envoy protos.
+        } catch (InvalidProtocolBufferException e) {
```

Comment:
I'd expect that we'd fail the conversion for both of these exceptions.

Reply:
Updated to fail, as we want to only consider the configuration of the first recognized policy.